File: src/Belajar_SOLID_06_OOP_Association/Association_02_Composition/Belanja.java
package Belajar_SOLID_06_OOP_Association.Association_02_Composition;

public class Belanja{
    public static void main(String[] args) {
        Pengguna user = new Pengguna("123","Cocoa",new Alamat("456","Surabaya"));
    }
}

class Pengguna{
    private String id;
    private String nama;
    private Alamat alamat;

    public Pengguna(String id,String nama ,Alamat alamat){
        this.id = id;
        this.alamat = alamat;
        this.nama = nama;
    }
}

class Alamat{
    private String id;
    private String location;

    public Alamat(String id ,String location){
        this.id = id;
        this.location = location;
    }
}
//    Perhatikan kode pada fungsi main() di atas, ketika kita membuat sebuah objek dari class User
//    maka kita juga membuat sebuah objek baru dari class Address. Apabila objek user kita hapus,
//    maka objek class address akan terhapus juga. Begitulah ilustrasi dari bentuk hubungan komposisi.

File: src/Belajar_SOLID_06_OOP_Association/Association_02_Composition/Belanja.kt
package Belajar_SOLID_06_OOP_Association.Association_02_Composition

// Komposisi = sebuah hubungan dapat dikatakan komposisi jika sebuah class memiliki class lainnya
// Aturan yang berlaku untuk hubungan ini adalah ketika kelas yang “memiliki” class
// lainnya dihancurkan maka class yang “dimiliki” oleh kelas tersebut akan hancur juga atau tidak dapat digunakan.

class User(
        private val id:String,
        private val name:String,
        private val address:Address
)

class Address(
        private val id:String,
        private val location:String
)

fun main(){
    val user = User(id = "1234",name = "Chino",address = Address("123","Yogyakarta"))
}

// Sebuah class User memiliki bentuk hubungan komposisi dengan
// class Address yang ditandai dengan adanya anak panah dengan ujung 
// diamond hitam penuh, yang mengarah pada class yang “memiliki” objek tersebut.
// Ketika class User hancur, maka class Address yang memiliki hubungan dengan
// class User tersebut akan ikut hancur dan tidak bisa digunakan lagi.